src/middlewared/middlewared/plugins/pool_/unlock.py

```diff
@@ -61,8 +61,8 @@ async def unlock_restarted_vms(self, dataset):
                 if (
                     (
                         dataset['type'] == 'FILESYSTEM' and
-                        (mountpoint := dataset_mountpoint(dataset)) and
-                        path.startswith(mountpoint + '/')
+                        ((mountpoint := dataset_mountpoint(dataset)) and path.startswith(mountpoint + '/')) or
+                        (path.startswith(zvol_name_to_path(dataset['name']) + '/'))
```

Comment:
This conditional isn't particularly readable. I feel like it should probably be broken apart into separate checks to make it easier to understand and maintain.

Shouldn't in this case be converting the zvol path to a dataset name and comparing the string with the dataset name being unlocked.

`/dev/zvol/tank/encrypted/my_zvol` -> `tank/encrypted/my_zvol` then compare that with what's my unlocked `tank/encrypted`